src/components/application_manager/include/application_manager/rpc_service_impl.h

```diff
@@ -0,0 +1,146 @@
+#include "formatters/formatter_json_rpc.h"
+#include "formatters/CFormatterJsonSDLRPCv2.h"
+#include "formatters/CFormatterJsonSDLRPCv1.h"
+#include "interfaces/HMI_API_schema.h"
+#include "interfaces/MOBILE_API_schema.h"
+
+#include "interfaces/v4_protocol_v1_2_no_extra.h"
+#include "interfaces/v4_protocol_v1_2_no_extra_schema.h"
+
+#include "utils/threads/message_loop_thread.h"
+#include "utils/logger.h"
+
+namespace application_manager {
+namespace rpc_service {
+
+namespace impl {
+
+struct MessageToMobile : public utils::SharedPtr<Message> {
+  MessageToMobile() : is_final(false) {}
+  explicit MessageToMobile(const utils::SharedPtr<Message>& message,
+                           bool final_message)
+      : utils::SharedPtr<Message>(message), is_final(final_message) {}
+  // PrioritizedQueue requres this method to decide which priority to assign
+  size_t PriorityOrder() const {
+    return (*this)->Priority().OrderingValue();
+  }
+  // Signals if connection to mobile must be closed after sending this message
+  bool is_final;
+};
+
+struct MessageToHmi : public utils::SharedPtr<Message> {
+  MessageToHmi() {}
+  explicit MessageToHmi(const utils::SharedPtr<Message>& message)
+      : utils::SharedPtr<Message>(message) {}
+  // PrioritizedQueue requres this method to decide which priority to assign
+  size_t PriorityOrder() const {
+    return (*this)->Priority().OrderingValue();
+  }
+};
+
+typedef threads::MessageLoopThread<utils::PrioritizedQueue<MessageToMobile> >
+    ToMobileQueue;
+typedef threads::MessageLoopThread<utils::PrioritizedQueue<MessageToHmi> >
+    ToHmiQueue;
+}
+
+class RPCServiceImpl : public RPCService,
+                       public impl::ToMobileQueue::Handler,
+                       public impl::ToHmiQueue::Handler {
+ public:
+  RPCServiceImpl(ApplicationManager& app_manager,
+                 request_controller::RequestController& request_ctrl,
+                 protocol_handler::ProtocolHandler* protocol_handler,
+                 hmi_message_handler::HMIMessageHandler* hmi_handler,
+                 CommandHolder& commands_holder);
+  ~RPCServiceImpl();
+
+  bool ManageMobileCommand(const commands::MessageSharedPtr message,
+                           commands::Command::CommandSource source) OVERRIDE;
+  bool ManageHMICommand(const commands::MessageSharedPtr message) OVERRIDE;
+
+  // CALLED ON messages_to_hmi_ thread!
+  void Handle(const impl::MessageToHmi message) OVERRIDE;
+  // CALLED ON messages_to_mobile_ thread!
+  void Handle(const impl::MessageToMobile message) OVERRIDE;
+
+  void SendMessageToMobile(const commands::MessageSharedPtr message,
+                           bool final_message = false) OVERRIDE;
+  void SendMessageToHMI(const commands::MessageSharedPtr message) OVERRIDE;
+
+  void set_protocol_handler(
+      protocol_handler::ProtocolHandler* handler) OVERRIDE;
+  void set_hmi_message_handler(
+      hmi_message_handler::HMIMessageHandler* handler) OVERRIDE;
+
+#ifdef SDL_REMOTE_CONTROL
```

Comment:
I think all RC related code is in rc_rpc_plugin, why do we have +#ifdef SDL_REMOTE_CONTROL ?